### ISSUE TITLE
Allow multiple versions of a single API to coexist on a single EWP Host

### ISSUE THREAD

**Issue by wrygiel** [erasmus-without-paper/ewp-specs-architecture#6]
A quote from the current specs:

> It is agreed that after an API is deployed onto production systems, all changes to this API SHOULD be kept backward-compatible - new major versions of such APIs SHOULD NOT be released. However, such APIs MAY become deprecated (and new APIs MAY be released in their place).

We should consider an alternative approach. If we allow multiple version of a single API to coexist on a single EWP Host, then we would still be able to introduce new major releases of deprecated APIs without breaking the system.
### Problem

Let's say we have API "X" in version `2.3.0`, and everyone is using it on their production systems.

A couple of years pass, and then suddenly a big change is required (e.g. a European law changes). Obviously we cannot expect everyone to upgrade to the updated version at the same time. We need to support both version of this API for some time (years, probably).
### Solution 1 (the current one)

The `1.0.0` architecture states that we cannot release a new version of the "X" API. We can however release a completely new API, let's call it "X2". This would be published in the manifest as below:

``` xml
<apis-implemented>
    <x version="2.3.0">...</x>
    <x2 version="1.0.0">...</x2>
</apis-implemented>
```

For some time after that, all EWP Hosts will support both APIs: the "X" API, and the "X2" API. Eventually, after majority of the clients have switched to the "X2" API, implementers may start to remove the "X" API from their EWP Hosts.

It's a great solution, it has just a single problem: the "2" suffix in the _name_ of the API is confusing. We could of course avoid having such numeric suffices, but perhaps there's a better solution.
### Solution 2 (the one we should consider)

If we allowed for multiple API versions to coexist on one server, then the clients would be able to choose the version which fits their needs:

``` xml
<apis-implemented>
    <x version="2.3.0">...</x>
    <x version="3.0.0">...</x>
</apis-implemented>
```

This looks better, but it introduces an extra requirement for the clients. Clients will be required to scan multiple `<x>` elements in search of the correct one, e.g. if the client has not yet adopted the `3.0.0` version then he must remember to look for the version `>=2.3.0,<3` (thanks to semantic versioning, all APIs which start with the same major number are always compatible with each other). Such dependency management seems to be a standard practice though (I would hope that most of the developers should have already seen or used the `>=2.3.0,<3` or `[2.3.0,3.0.0)` syntax in their apps).

**Comment by wrygiel** [comment]
@Awerin

**Comment by mkurzydlowski** [comment]
Does solution 2 mean that the client needs to communicate to the server which version of the method he wants to use for the current request?

**Comment by wrygiel** [comment]
No, both versions would need to have a different URL. E.g.

``` xml
<apis-implemented>
    <x version="2.3.0">
        <url>The URL at which version 2.3.0 is implemented.</url>
    </x>
    <x version="3.0.0">
        <url>The URL at which version 3.0.0 is implemented.</url>
    </x>
</apis-implemented>
```

**Comment by wrygiel** [comment]
Released in Discovery Manifest API v2.0.0.
